gems/sorbet-runtime/lib/types/props/private/initializer_generator.rb

```diff
@@ -0,0 +1,183 @@
+      end
+      def self.generate(props, weak:)
+        parts = props.map do |prop, rules|
+          # All of these strings should already be validated (directly or
+          # indirectly) in `validate_prop_name`, so we don't bother with a nice
+          # error message, but we double check here to prevent a refactoring
+          # from introducing a security vulnerability.
+          raise unless T::Props::Decorator::SAFE_NAME.match?(prop.to_s)
+
+          ivar_name = rules.fetch(:accessor_key).to_s
+          raise unless ivar_name.start_with?('@') && T::Props::Decorator::SAFE_NAME.match?(ivar_name[1..-1])
+
+          # It seems like a bug that this affects the behavior of setters, but
+          # some existing code relies on this behavior
+          has_explicit_nil_default = rules.key?(:default) && rules.fetch(:default).nil?
+          need_nil_write_check = T::Props::Utils.need_nil_write_check?(rules) && !has_explicit_nil_default
+
+          missing_handler, needs_missing_typecheck = generate_missing_handler(
+            prop: prop,
+            rules: rules,
+            ivar_name: ivar_name,
+            need_nil_write_check: need_nil_write_check,
+            weak: weak,
+          )
+
+          typecheck_handler = generate_typecheck_handler(
+            prop: prop,
+            rules: rules,
+            ivar_name: ivar_name,
+            need_nil_write_check: need_nil_write_check,
+          )
+
+          <<~RUBY.strip
+            val = hash[#{prop.inspect}]
+            if val.nil? && !hash.key?(#{prop.inspect})
+              found -= 1
+              #{ivar_name} = #{missing_handler}
+              #{needs_missing_typecheck ? typecheck_handler : ''}
+            else
+              #{ivar_name} = val
+              #{typecheck_handler}
+            end
+          RUBY
+        end
+
+        <<~RUBY
+          def __t_props_generated_initialize(hash)
+            found = #{props.size}
+            decorator = self.class.decorator
+
+            #{parts.join("\n\n")}
+
+            if found < hash.size
+              raise ArgumentError.new("\#{self.class}: Unrecognized properties: \#{(hash.keys - decorator.props.keys).join(', ')}")
+            end
+          end
+        RUBY
+      end
+
+      sig do
+        params(
+          prop: Symbol,
+          rules: T::Hash[Symbol, T.untyped],
+          ivar_name: String,
+          need_nil_write_check: T::Boolean,
+          weak: T::Boolean,
+        )
+        .returns([String, T::Boolean])
+        .checked(:never)
+      end
+      private_class_method def self.generate_missing_handler(prop:, rules:, ivar_name:, need_nil_write_check:, weak:)
+        if rules.key?(:factory)
+          ["self.class.class_exec(&#{decorator_fetch(prop, :factory)})", true]
+        elsif rules.key?(:default)
+          default = rules.fetch(:default)
+
+          literal_default = case default
+          when Integer, Symbol, Float, TrueClass, FalseClass, NilClass
```

Comment:
Why is this not doing something more like what `generate_nil_handler` for deserialization does?  Seems like adopting that structure would enable you to deal more elegantly with e.g arrays and hashes.  I realize that for certain things, going through `self.class.decorator.props_with_defaults.fetch(#{prop.inspect}).default` would probably be slower than what you have here (e.g. `T::Enum`), but it would probably be more correct?